lib/auth/webauthn/login.go

```diff
@@ -0,0 +1,289 @@
+	storedUser, err := f.Identity.GetUser(user, true /* withSecrets */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	u := newWebUser(storedUser, false /* idOnly */, []*types.MFADevice{dev})
+
+	// Fetch the previously-stored SessionData, so it's checked against the user
+	// response.
+	sessionDataPB, err := f.Identity.GetWebAuthnSessionData(user, loginSessionID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	sessionData := sessionFromPB(sessionDataPB)
+
+	// Create a WebAuthn matching the expected RPID and Origin, then verify the
+	// signed challenge.
+	web, err := newWebAuthn(f.Webauthn, rpID, origin)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credential, err := web.ValidateLogin(u, *sessionData, parsedResp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Update last used timestamp and device counter.
+	if err := setCounterAndTimestamps(dev, credential); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := f.Identity.UpsertMFADevice(ctx, user, dev); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The user just solved this challenge, so let's make sure it won't be used
+	// again.
+	if err := f.Identity.DeleteWebAuthnSessionData(user, loginSessionID); err != nil {
+		log.Warnf("WebAuthn: failed to delete SessionData for user %v", user)
+	}
+
+	return dev, nil
+}
+
+func parseCredentialResponse(resp *CredentialAssertionResponse) (*protocol.ParsedCredentialAssertionData, error) {
+	// Remove extensions before Marshal, they are not supported.
+	exts := resp.PublicKeyCredential.Extensions
+	resp.PublicKeyCredential.Extensions = nil
+	defer func() { resp.PublicKeyCredential.Extensions = exts }()
+
+	// This is a roundabout way of getting resp validated, but unfortunately the
+	// APIs don't provide a better method (and it seems better than duplicating
+	// library code here).
+	body, err := json.Marshal(resp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return protocol.ParseCredentialRequestResponseBody(bytes.NewReader(body))
+}
+
+func validateOrigin(origin, rpID string) error {
+	parsedOrigin, err := url.Parse(origin)
+	if err != nil {
+		return trace.BadParameter("origin is not a valid URL: %v", err)
+	}
+	host, err := utils.Host(parsedOrigin.Host)
+	if err != nil {
+		return trace.BadParameter("extracting host from origin: %v", err)
+	}
+
+	// TODO(codingllama): Check origin against the public addresses of Proxies and
+	//  Auth Servers
+
+	// Accept origins whose host matches the RPID.
+	if host == rpID {
+		return nil
+	}
+
+	// Accept origins whose host is a subdomain of RPID.
+	originParts := strings.Split(host, ".")
+	rpParts := strings.Split(rpID, ".")
+	if len(originParts) <= len(rpParts) {
+		return trace.BadParameter("origin doesn't match RPID")
+	}
+	i := len(originParts) - 1
+	j := len(rpParts) - 1
+	for j >= 0 {
+		if originParts[i] != rpParts[j] {
+			return trace.BadParameter("origin doesn't match RPID")
+		}
+		i--
+		j--
+	}
```

Comment:
For cleanliness, we can probably separate this out into a different function: `OriginMatchesRPID(origin, rpid)`

Reply:
Since this is much of what validateOrigin does, I'll keep as is for now (but will refactor later if this keeps growing).